src/stim/io/measure_record_reader.cc

```diff
@@ -415,124 +419,95 @@ void MeasureRecordReaderFormatR8::buffer_data() {
     } else if (total_data > bits_per_record + 1) {
         throw std::invalid_argument("r8 data encoded a jump past the expected end of encoded data.");
     }
+    return true;
 }
 
 /// DETS format
 
 MeasureRecordReaderFormatDets::MeasureRecordReaderFormatDets(
     FILE *in, size_t n_measurements, size_t n_detection_events, size_t n_logical_observables)
     : in(in),
-      bits_per_m_record(n_measurements),
-      bits_per_d_record(n_detection_events),
-      bits_per_l_record(n_logical_observables) {
-    if (!maybe_consume_keyword(in, "shot", separator)) {
-        throw std::runtime_error("Need a \"shot\" to begin record");
-    }
-    update_next_shot();
-    result_type = next_shot_result_type;
+      buffer(n_measurements + n_detection_events + n_logical_observables),
```

Comment:
I would prefer to follow the robustness principle of "be liberal in what you accept" on this one. Also notice how I allow entries to be repeated, which could plausibly happen if a writer was repeating detection events triggered from multiple errors, instead of cancelling them out.

One major reason I think ordering is a bit tricky is that, in the data format, logical observables come at the end. But during simulations observables may come and go at intermediate times. And so they may be written much earlier than some of the detectors.

I do think it would be reasonable to have a boolean flag to switch to a stricter mode that assumed ordered to enable streamability.